sql-plugin/src/main/330db/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastHashJoinExec.scala

```diff
@@ -0,0 +1,226 @@
+  def shuffleExchange: GpuShuffleExchangeExec = buildPlan match {
+    case bqse: ShuffleQueryStageExec if bqse.plan.isInstanceOf[GpuShuffleExchangeExec] =>
+      bqse.plan.asInstanceOf[GpuShuffleExchangeExec]
+    case bqse: ShuffleQueryStageExec if bqse.plan.isInstanceOf[ReusedExchangeExec] =>
+      bqse.plan.asInstanceOf[ReusedExchangeExec].child.asInstanceOf[GpuShuffleExchangeExec]
+    case gpu: GpuShuffleExchangeExec => gpu
+    case reused: ReusedExchangeExec => reused.child.asInstanceOf[GpuShuffleExchangeExec]
+  }
+
+
+  private def getExecutorBuiltBatchAndStreamIter(
+      buildRelation: RDD[ColumnarBatch],
+      buildSchema: StructType,
+      buildOutput: Seq[Attribute],
+      streamIter: Iterator[ColumnarBatch],
+      coalesceMetricsMap: Map[String, GpuMetric]): (ColumnarBatch, Iterator[ColumnarBatch]) = {
+    val semWait = coalesceMetricsMap(GpuMetric.SEMAPHORE_WAIT_TIME)
+    val targetSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(conf)
+    val metricsMap = allMetrics
+
+    val bufferedStreamIter = new CloseableBufferedIterator(streamIter.buffered)
+    closeOnExcept(bufferedStreamIter) { _ =>
+      withResource(new NvtxRange("first stream batch", NvtxColor.RED)) { _ =>
+        if (bufferedStreamIter.hasNext) {
+          bufferedStreamIter.head
+        } else {
+          GpuSemaphore.acquireIfNecessary(TaskContext.get(), semWait)
+        }
+      }
+      val buildBatch = GpuExecutorBroadcastHelper.getExecutorBroadcastBatch(buildRelation,
+          buildSchema, buildOutput, metricsMap, targetSize)
+      (buildBatch, bufferedStreamIter)
+    }
+  }
+
+  private def doExecutorBroadcastJoin(): RDD[ColumnarBatch] = {
+    val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
+    val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
+    val opTime = gpuLongMetric(OP_TIME)
+    val streamTime = gpuLongMetric(STREAM_TIME)
+    val joinTime = gpuLongMetric(JOIN_TIME)
+    val joinOutputRows = gpuLongMetric(JOIN_OUTPUT_ROWS)
+
+    val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
+
+    val targetSize = RapidsConf.GPU_BATCH_SIZE_BYTES.get(conf)
+
+    val buildRelation = shuffleExchange.executeColumnar()
+
+    val rdd = streamedPlan.executeColumnar()
+    val localBuildSchema = buildPlan.schema
+    val localBuildOutput = buildPlan.output
+    rdd.mapPartitions { it =>
+      val (builtBatch, streamIter) =
+        getExecutorBuiltBatchAndStreamIter(
+          buildRelation,
+          localBuildSchema,
+          localBuildOutput,
+          new CollectTimeIterator("executor broadcast join stream", it, streamTime),
+          allMetrics)
+      withResource(builtBatch) { _ =>
+        doJoin(builtBatch, streamIter, targetSize, spillCallback,
+          numOutputRows, joinOutputRows, numOutputBatches, opTime, joinTime)
+      }
+    }
+  }
+
+  override def doExecuteColumnar(): RDD[ColumnarBatch] = {
+    if (isExecutorBroadcast) {
+      doExecutorBroadcastJoin()
+    } else {
+      doColumnarBroadcastJoin()
```

Comment:
`doExecutorBroadcastJoin` vs `doColumnarBroadcastJoin` feels a bit confusing. Both are columnar.

This is a minor point, it's hard to come up with a good name. Perhaps `doExecutorBroadcastJoin` -> `doExecutorColumnarBatchJoin`?